subsys/mgmt/mcumgr/lib/cmd/img_mgmt/src/zephyr_img_mgmt.c

```diff
@@ -338,30 +338,33 @@ img_mgmt_impl_read(int slot, unsigned int offset, void *dst,
  * The alloc_ctx and free_ctx are specifically provided for
  * the img_mgmt_impl_write_image_data to allocate/free single flash_img_context
  * type buffer.
- * When heap is enabled these functions will operate on heap; when  heap is not
- * allocated the alloc_ctx just returns pointer to static, global life-time
- * variable, and free_ctx does nothing.
- * CONFIG_HEAP_MEM_POOL_SIZE is C preprocessor literal.
  */
 static inline struct flash_img_context *alloc_ctx(void)
 {
-	struct flash_img_context *ctx = NULL;
+#if defined(CONFIG_IMG_MGMT_USE_HEAP_FOR_FLASH_IMG_CONTEXT)
+	/* Even CONFIG_HEAP_MEM_POOL_SIZE will be able to match size of the structure,
+	 * keep in mind that when application will put the heap under pressure, obtaining
+	 * of flash image context may not be possible, so plan bigger heap size or
+	 * make sure to limit application pressure on heap when DFU is expected.
+	 */
+	BUILD_ASSERT(CONFIG_HEAP_MEM_POOL_SIZE >= (sizeof(struct flash_img_context)),
+		     "Not enough heap mem for flash_img_context.");
 
-	if (CONFIG_HEAP_MEM_POOL_SIZE > 0) {
-		ctx = k_malloc(sizeof(*ctx));
-	} else {
-		static struct flash_img_context stcx;
+	return k_malloc(sizeof(struct flash_img_context));
+#else
+	static struct flash_img_context stcx;
 
-		ctx = &stcx;
-	}
-	return ctx;
+	return &stcx;
```

Comment:
Would it be crazy to allocate from a memory block here instead, with just enough size for a single `flash_img_context`? Then the `free` function could do something useful, at least. If this has always been meant to allocate at most one such value, I don't really get why this is using the alloc/free function naming pattern.

Reply:
> Would it be crazy to allocate from a memory block here instead, with just enough size for a single `flash_img_context`? Then the `free` function could do something useful, at least. If this has always been meant to allocate at most one such value, I don't really get why this is using the alloc/free function naming pattern.

I think that original idea was that the `flash_img_context` may be significant in size, as it (`flash_img_context`) contains a cached data buffer controlled by the `CONFIG_IMG_BLOCK_BUF_SIZE`, so it may take away a significant size of the RAM, when allocated statically. For example when you define the config to be 4k (it is 0.5k by default), you take away more than 4k from the start of an application.
Somebody, probably, thought that it may be worth to use the heap for such a context as it will take the RAM only for a duration of DFU, so most of application runtime the RAM will be available for other tasks.
Unfortunately, this was not left for the user to choose and an average user would not even know that the heap, they have defined, will be taken for such purpose, and may be surprised when suddenly DFU would stop working.